server/src/main/java/io/crate/execution/engine/collect/LuceneShardCollectorProvider.java

```diff
@@ -119,10 +119,9 @@ protected BatchIterator<Row> getUnorderedIterator(RoutedCollectPhase collectPhas
         SharedShardContext sharedShardContext = collectTask.sharedShardContexts().getOrCreateContext(shardId);
         var searcher = sharedShardContext.acquireSearcher("unordered-iterator: " + formatSource(collectPhase));
         collectTask.addSearcher(sharedShardContext.readerId(), searcher);
-        IndexShard sharedShardContextShard = sharedShardContext.indexShard();
```

Comment:
I wonder if existing `IndexShard` and one we get from the `SharedShardContext` are always the same.

Checking `SharedShardContext getOrCreateContext()`, it looks like if we don't find in cache, we end up resolving it via id: `indexService.getShard(shardId.id()) `, so should be same?

Reply:
They should be the same, unless there's a race where a shard gets removed & created on the same node within the window. But that likely causes failures & retries anyway and was a problem before too. The change here might even mitigate it somewhat.